js/compiler_tests.ts

```diff
@@ -0,0 +1,466 @@
+  }
+};
+
+const getEmitOutputPassThrough = true;
+
+const serviceMock = {
+  getCompilerOptionsDiagnostics(): ts.Diagnostic[] {
+    return originals._service.getCompilerOptionsDiagnostics.call(
+      originals._service
+    );
+  },
+  getEmitOutput(fileName: string): ts.EmitOutput {
+    getEmitOutputStack.push(fileName);
+    if (getEmitOutputPassThrough) {
+      return originals._service.getEmitOutput.call(
+        originals._service,
+        fileName
+      );
+    }
+    if (fileName in emittedFiles) {
+      return {
+        outputFiles: [{ text: emittedFiles[fileName] }] as any,
+        emitSkipped: false
+      };
+    }
+    return { outputFiles: [], emitSkipped: false };
+  },
+  getSemanticDiagnostics(fileName: string): ts.Diagnostic[] {
+    return originals._service.getSemanticDiagnostics.call(
+      originals._service,
+      fileName
+    );
+  },
+  getSyntacticDiagnostics(fileName: string): ts.Diagnostic[] {
+    return originals._service.getSyntacticDiagnostics.call(
+      originals._service,
+      fileName
+    );
+  }
+};
+const windowMock: { define?: compiler.AmdDefine } = {};
+const mocks = {
+  _globalEval: globalEvalMock,
+  _log: logMock,
+  _os: osMock,
+  _ts: tsMock,
+  _service: serviceMock,
+  _window: windowMock
+};
+
+// Setup the mocks
+test(function compilerTestsSetup() {
+  assert("_globalEval" in compilerInstance);
+  assert("_log" in compilerInstance);
+  assert("_os" in compilerInstance);
+  assert("_ts" in compilerInstance);
+  assert("_service" in compilerInstance);
+  assert("_window" in compilerInstance);
+  Object.assign(compilerInstance, mocks);
+});
+
+test(function compilerInstance() {
+  assert(DenoCompiler != null);
+  assert(DenoCompiler.instance() != null);
+});
+
+// Testing the internal APIs
+
+test(function compilerMakeDefine() {
+  const moduleMetaData = new ModuleMetaData(
+    "/root/project/foo/bar.ts",
+    fooBarTsSource,
+    fooBarTsOutput
+  );
+  const localDefine = compilerInstance.makeDefine(moduleMetaData);
+  let factoryCalled = false;
+  localDefine(
+    ["require", "exports", "compiler"],
+    (_require, _exports, _compiler): void => {
+      factoryCalled = true;
+      assertEqual(
+        typeof _require,
+        "function",
+        "localRequire should be a function"
+      );
+      assert(_exports != null);
+      assert(
+        Object.keys(_exports).length === 0,
+        "exports should have no properties"
+      );
+      assert(compiler === _compiler, "compiler should be passed to factory");
+    }
+  );
+  assert(factoryCalled, "Factory expected to be called");
+});
+
+// TODO testMakeDefineExternalModule - testing that make define properly runs
+// external modules, this is implicitly tested though in
+// `compilerRunMultiModule`
+
+test(function compilerRun() {
+  // equal to `deno foo/bar.ts`
+  const codeFetchLength = codeFetchStack.length;
+  const codeCacheLength = codeCacheStack.length;
+  const globalEvalLength = globalEvalStack.length;
```

Comment:
Awesome work on the tests. Why not just assert that these are empty here, rather than recording the length? Since you're always popping off everything, they should be empty.

Reply:
Done